docs/rfcs/020-pageserver-s3-coordination.md

```diff
@@ -0,0 +1,75 @@
+
+### Standby pageserver
+
+Another related case is standby pageserver. In this case second pageserver can be used as a replica to scale reads and serve as a failover target in case first one fails.
+
+In this mode second pageserver needs to have the same picture of s3 files to be able to load layers on-demand. To accomplish that second pageserver cannot run gc/compaction jobs. Instead it needs to receive updates for index contents. (There is no need to run walreceiver on the second pageserver then).
+
+## Observations
+
+- If both pageservers ingest wal then their layer set diverges, because layer file generation is not deterministic
+- If one of the pageservers does not ingest wal (and just picks up layer updates) then it lags behind and cannot really answer queries in the same pace as the primary one
+- Can compaction help make layers deterministic? E g we do not upload level zero layers and construction of higher levels should be deterministic. This way we can guarantee that layer creation by timeout wont mess things up. This way one pageserver uploads data and second one can just ingest it. But we still need some form of election
+
+## Solutions
+
+### Manual orchestration
+
+One possible solution for relocation case is to orchestrate background jobs from outside. The oracle who runs migration can turn off background jobs on PS1 before migration and then run migration -> enable them on PS2. The problem comes if migration fails. In this case in order to resume background jobs oracle needs to guarantee that PS2 doesnt run background jobs and if it doesnt respond then PS1 is stuck unable to run compaction/gc. This cannot be solved without human ensuring that no upload from PS2 can happen. In order to be able to resolve this automatically CAS is required on S3 side so pageserver can avoid overwriting index part if it is no longer the leading one
+
+Note that flag that disables background jobs needs to be persistent, because otherwise pageserver restart will clean it
+
+### Avoid index_part.json
+
+Index part consists of two parts, list of layers and metadata. List of layers can be easily obtained by \`ListObjects\` S3 API method. But what to do with metadata? Create metadata instance for each checkpoint and add some counter to the file name?
+
+Back to potentially long s3 ls.
+
+### Coordination based approach
+
+Do it like safekeepers chose leader for WAL upload. Ping each other and decide based on some heuristics e g smallest node id. During relocation PS1 sends "resign" ping message so others can start election without waiting for a timeout.
+
+This still leaves metadata question open and non deterministic layers are a problem as well
```

Comment:
I like this approach more, since we do something similar with SKs already and SKs could benefit from the information pageservers post to etcd (shared dashboard).

In addition, it appears to me that etcd allows to solve the metadata sync problem:
> If one of the pageservers does not ingest wal (and just picks up layer updates) then it lags behind and cannot really answer queries in the same pace as the primary one

could be easily solved if the timeline leader pushes index_part.json contents as a ping message: this way every other PS that's interested in the timeline but not a leader, can subscribe and get the fresh data.

On leader reelection, we can reread the index file and presumably also keep the most up-to-date info about the remote layers on the node side that writes to S3.

Reply:
> could be easily solved if the timeline leader pushes index_part.json contents as a ping message

I think this is not the case. When layer is generated is it already too late. By default layer is generated after each 256 megabytes of WAL. If we wait for timeout it is even worse.

Reply:
Ah, so the issue is that we don't share the non-flushed data with the other pageservers that serve the same get_page requests for the same timeline?

Similar to SK nodes, we could coordinate the connection strings in our blackboard messages and proxy the get_page requests to other nodes, if they have the Lsn the current node does not have yet.

Reply:
> and proxy the get_page requests to other nodes, if they have the Lsn the current node does not have yet.

This makes sense.

In general we would then have the problem of not having those newer WAL segments in case the leader disappears, but we already have the safekeepers which would be able to assist with the recovery from the latest flushed layer file onwards.

Reply:
> Ah, so the issue is that we don't share the non-flushed data with the other pageservers that serve the same get_page requests for the same timeline?

Correct

> proxy the get_page requests to other nodes

I see two downsides there. 

1. Failover process
2. Uneven load distribution 

Let me go into details for each point:
1. If primary dies than replica wont be able to forward such requests. In order to continue operations it needs to detect that primary has failed (2 ping timeouts or something). Then it needs to start replication, and then finally ingest WAL needed to answer `get_page` request. If both pageservers ingest WAL most probably they do that in the same pace, so  in this case there are no actions needed for failover from pageserver side. Compute can detect failed pageserver and route queries to the healthy one. Or as aurora does request pages from both and pick the first reply
2. With this forwarding primary will be always responsible for both, ingesting wal (writes) an for reads of the recent data. I think may not help to scale reads unless significant portion of queries goes into old data (this may be the case for read replica postgres), because replica wont have data for tip of the timeline so for such queries it will only add forwarding latency without any performance boost

Reply:
For 1., I would imagine that we keep all nodes as up-to-date as possible for the timeline, potentially evicting some _old_ (LRU?) layers based on some heuristics. 

Hence,

> so in this case there are no actions needed for failover from pageserver side

imagined that would be mostly the case.

Reply:
1. Didnt understand your point. With this "ping on upload" we'll have 256+ mb lag. Eviction especially for old layers shouldnt interfere with this. So the problem persists.
2. If layer generation is deterministic then there is no need to perform s3 ls other than during startup. Replica will ingest wal, create local layers and on eviction just drop them so they're still in layer map as remote ones

Reply:
1. Oh, no, no "ping on upload" but rather ping every N (1) seconds as SK do, just update the data between pings, if needed.
So, we would have a more or less up-to-date vision on PS "cluster" state around the timeline and who's the leader.

Reply:
1. Do not understand how it can solve lag problem. You mean replicating timeline data through broker?

But otherwise yes. SK-like model is applicable to sync cluster state